basic_uint.h

```diff
@@ -286,11 +287,51 @@ namespace mgt {
     }
 
     friend basic_uint<l> operator*(const basic_uint<l> &a, const basic_uint<l> &b) {
-      basic_uint<l> res, tmp = b;
-      for (size_t i = 0; i < l; ++i, tmp <<= 1) {
-        if (a.data[i])
-          res += tmp;
+      basic_uint<l> res(0);
+      int current_numa, current_numb;
+      int temp_res;
+      std::vector<int> hexa, hexb;
```

Comment:
if the size of the vector is predictable, try to use member function `reserve(size_t size)` to avoid unnecessary memory allocation.